### PR TITLE
Expand functionality of message channels by introducing a Channel ID

### PR DESCRIPTION
**Pull request by dmitry-ganzha**
This PR introduces the `ChannelId` for the `MessageChannel`, the utility `Channels` for working with message channels and the `Router` for routing messages to the corresponding message channels.

**Highlights:**

- The `ChannelId` is the identifier of a channel based on some grouping. For instance, it can be `TypeUrl` allows grouping messages by the same message type.

- `Channels` is the utility class for working with message channels and at the moment can generate channel identifiers for various objects for example for `MessageClass`.

-  The `Router` is the base interface for routers. The router is designed for routing a message based on some conditions which can be updated in runtime. Inspired by [Dynamic Router pattern](http://www.enterpriseintegrationpatterns.com/patterns/messaging/DynamicRouter.html).

-  The `Route` is the base interface for routes. The `Route` is designed for representing the connection between the router and the message channel and checking if the message is acceptable for the route or not via `MessageRouted accept(ExternalMessage message)`.

Prior to these changes, the message channel was strictly tied to the type of messages it can pass. There are places in Spine in which messages should be grouped not only according to the message type. Therefore, a `ChannelId` has been introduced to solve this problem. At the moment `ChannelId` consists only of one possible grouping identifier, the **message_type_url**. It can be used to group messages in a channel by the `MessageClass`. 

The method `Channels.newId(MessageClass messageClass)` can be used to generate the `ChannelId` related to the `MessageClass`.

In the nearest future will be added more channel identifiers for other needs.

Also, in the scope of this PR was refactored the `IntegrationBus` which uses message channels grouped by a message class.

The framework version is set to 0.10.4-SNAPSHOT.

### CODE REVIEW

**Comment by dmitry-ganzha** [comment]
@armiol PTAL

File: server/src/main/proto/spine/server/integration/message_channel.proto [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+option java_package = "io.spine.server.integration";
+option java_multiple_files = true;
+option java_outer_classname = "MessageChannelProto";
+option java_generate_equals_and_hash = true;
+
+// The message which is used to identify message channel.
+message ChannelId {
+
+    // Identifier of message channel, it used for grouping messages by various characteristics.
+    oneof channel_id {
```
Comment (armiol) [inline server/src/main/proto/spine/server/integration/message_channel.proto]: Its better not be named in the same way as its parent. Right now it's `ChannelId.channel_id`, which is weird.

I'd propose something like `kind`, which is a generic term we use in this kind of situation (see other `oneof`s).

File: server/src/main/proto/spine/server/integration/message_channel.proto [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import "spine/options.proto";
+
+option (type_url_prefix) = "type.spine.io";
+option java_package = "io.spine.server.integration";
+option java_multiple_files = true;
+option java_outer_classname = "MessageChannelProto";
+option java_generate_equals_and_hash = true;
+
+// The message which is used to identify message channel.
```
Comment (armiol) [inline server/src/main/proto/spine/server/integration/message_channel.proto]: Either an article is missing, or it should be `...is used to identify message channels.`

File: server/src/main/proto/spine/server/integration/message_channel.proto [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+message ChannelId {
+
+    // Identifier of message channel, it used for grouping messages by various characteristics.
```
Comment (armiol) [inline server/src/main/proto/spine/server/integration/message_channel.proto]: I'd say it's a channel kind (see more on that below).

And in this context we are identifying (not grouping) the messages by their kind.

File: server/src/test/java/io/spine/server/integration/ChannelsShould.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import io.spine.core.EventClass;
+import org.junit.Test;
+
+import static io.spine.test.Tests.assertHasPrivateParameterlessCtor;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * @author Dmitry Ganzha
+ */
+public class ChannelsShould {
```
Comment (armiol) [inline server/src/test/java/io/spine/server/integration/ChannelsShould.java]: Can we please check the designed behavior of the methods as well?

File: server/src/main/java/io/spine/server/integration/MessageChannel.java [hunk 5]
```diff
@@ -20,20 +20,19 @@
 package io.spine.server.integration;
 
 /**
- * A channel dedicated to exchanging the messages of a single message type.
- *
- * <p>The class of message serves as a channel key.
+ * A channel dedicated to exchanging the messages.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/MessageChannel.java]: Please think of something else. This description is duplicating the class title.

File: server/src/main/java/io/spine/server/integration/IntegrationBus.java [hunk 6]
```diff
@@ -262,14 +265,16 @@ public void register(ExternalMessageDispatcher<?> dispatcher) {
...
 
-        // Subscribe to incoming messages of requested types.
+        // Unsubscribe from incoming messages from the requested channels.
         unsubscribeFromIncoming(dispatcher);
 
-        final Set<ExternalMessageClass> currentlyRequested = subscriberHub.keys();
+        final Set<ChannelId> currentlyRequested = subscriberHub.keys();
         if (!currentlyRequested.equals(requestedBefore)) {
+
+            // Notify others that the requested channels has been changed.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/IntegrationBus.java]: Please fix the grammar issues.

File: server/src/main/java/io/spine/server/integration/IntegrationBus.java [hunk 7]
```diff
@@ -238,22 +241,22 @@ protected void store(Iterable<ExternalMessage> messages) {
...
-        // Subscribe to incoming messages of requested types.
+        // Subscribe to incoming messages from the requested channels.
         subscribeToIncoming(dispatcher);
 
-        final Set<ExternalMessageClass> currentlyRequested = subscriberHub.keys();
+        final Set<ChannelId> currentlyRequested = subscriberHub.keys();
         if (!currentlyRequested.equals(requestedBefore)) {
 
-            // Notify others that the requested message set has been changed.
+            // Notify others that the requested channels has been changed.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/IntegrationBus.java]: Please fix the grammar.

File: server/src/main/java/io/spine/server/integration/IntegrationBus.java [hunk 8]
```diff
@@ -238,22 +241,22 @@ protected void store(Iterable<ExternalMessage> messages) {
     public void register(ExternalMessageDispatcher<?> dispatcher) {
         super.register(dispatcher);
 
-        // Remember the message types, that we have been subscribed before.
-        final Set<ExternalMessageClass> requestedBefore = subscriberHub.keys();
+        // Remember the channels, that we have been subscribed before.
+        final Set<ChannelId> requestedBefore = subscriberHub.keys();
 
-        // Subscribe to incoming messages of requested types.
+        // Subscribe to incoming messages from the requested channels.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/IntegrationBus.java]: The problem with the fixes in the comments is that `IntegrationBus` subscribes to messages. And sends requests to other parties (such as remote `IntegrationBus` instances) to tell them that the messages of certain types are needed.

The changes made to the comments mostly destroy this meaning. Now, for some reason, `IntegrationBus` subscribes to some channels, and that's it. Totally unclear to me.

File: server/src/main/java/io/spine/server/integration/TransportFactory.java [hunk 9]
```diff
@@ -30,25 +29,24 @@
...
  *
  * @author Alex Tymchenko
+ * @author Dmitry Ganzha
  */
 @SPI
 public interface TransportFactory {
 
     /**
-     * Creates a {@link Publisher} for the messages of given class.
+     * Creates a {@link Publisher} for the messages of given channel identifier.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/TransportFactory.java]: `for the messages of given channel identifier.` 

Here and below does not make much sense. Messages cannot be of a `channel identifier`.

**Comment by dmitry-ganzha** [comment]
@armiol PTAL

File: server/src/main/java/io/spine/server/integration/Channels.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.integration;
+
+import io.spine.type.MessageClass;
+import io.spine.type.TypeUrl;
+
+/**
+ * Utilities for working with message channels.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/Channels.java]: Please have a `{@linkplain MessageChannel message channels}`.

File: server/src/main/java/io/spine/server/integration/Channels.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * @author Dmitry Ganzha
+ */
+public final class Channels {
+
+    /** Prevents instantiation on this utility class. */
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/Channels.java]: `of this utility class.`

File: server/src/main/java/io/spine/server/integration/IntegrationBus.java [hunk 12]
```diff
@@ -262,14 +267,18 @@ public void register(ExternalMessageDispatcher<?> dispatcher) {
     public void unregister(ExternalMessageDispatcher<?> dispatcher) {
         super.unregister(dispatcher);
 
-        // Remember the message types, that we have been subscribed before.
-        final Set<ExternalMessageClass> requestedBefore = subscriberHub.keys();
+        // Remember the channels, that we have been subscribed before.
+        // The channel ID contains a type of messages which can be passed through the
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/IntegrationBus.java]:
```
// The channel ID corresponds to the type of messages, 
// that are being served through the channel.
```

File: server/src/main/java/io/spine/server/integration/IntegrationBus.java [hunk 13]
```diff
@@ -282,26 +291,27 @@ public void unregister(ExternalMessageDispatcher<?> dispatcher) {
      * request for external messages} for that purpose.
      *
      * @param  currentlyRequested
-     *         the set of message types that are now requested by this instance of
-     *         integration bus
+     *         the set of channel identifiers containing message types that are now requested
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/IntegrationBus.java]: `the set of channel IDs, which correspond to the currently requested types of messages.`

File: server/src/main/java/io/spine/server/integration/MessageChannel.java [hunk 14]
```diff
@@ -20,20 +20,21 @@
 package io.spine.server.integration;
 
 /**
- * A channel dedicated to exchanging the messages of a single message type.
+ * A channel dedicated to exchanging the messages which grouped by some rule.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/MessageChannel.java]:
```
A channel for exchanging the {@link com.google.protobuf.Message Message}s.

<p>Identified by a {@linkplain ChannelId channel ID}, which must be unique in scope of the underlying messaging system.

<p>The identifier is also used to specify the set of {@code com.google.protobuf.Message}s, suitable for this channel. The definition of the matching criterion is done according to the {@linkplain ChannelId#getKind() kind} value.

@see ChannelId ...
...
```

File: server/src/main/java/io/spine/server/integration/TransportFactory.java [hunk 15]
```diff
@@ -30,25 +29,26 @@
...
  *
  * @author Alex Tymchenko
+ * @author Dmitry Ganzha
  */
 @SPI
 public interface TransportFactory {
 
     /**
-     * Creates a {@link Publisher} for the messages of given class.
+     * Creates a {@link Publisher} for the messages that belong to the kind
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/TransportFactory.java]:
```
Creates a {@link Publisher} for the messages that correspond to the passed channel ID.
```

We also need to have some abstraction, that tells whether the message matches the given channel ID or not. It should probably be implemented in scope of `Channels`. Then we could reference it in this description (e.g.  "correspond" could be a "@linkplain" to that method). We should probably invoke it during the publishing.

Please do that in scope of this PR.  Let's discuss if needed.

**Comment by dmitry-ganzha** [comment]
@armiol PTAL, also I would like to discuss the command MessageSuitableByMessageType.

**Comment by dmitry-ganzha** [comment]
@armiol PTAL.

**Comment by dmitry-ganzha** [comment]
@armiol PTAL. I still need to finish some tests but I would like to get feedback on the existing code.

**Comment by armiol** [comment]
@dmitry-ganzha please complete the tests before the review starts. Feel free to notify me once it's done.

**Comment by dmitry-ganzha** [comment]
@armiol PTAL.

File: server/src/main/java/io/spine/server/integration/Channels.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    /** Prevents instantiation of this utility class. */
+    private Channels() {}
+
+    /**
+     * Generates the {@code ChannelId} for the passed {@code MessageClass}.
+     *
+     * @param messageClass a message class for which a channel identifier is generated
+     * @return a channel identifier
+     */
+    public static ChannelId newId(MessageClass messageClass) {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/Channels.java]: What if this is `null`? Please see how other utilities are tested.

File: server/src/main/java/io/spine/server/integration/DomesticRejectionPublisher.java [hunk 17]
```diff
@@ -66,12 +73,18 @@
     public Set<String> dispatch(RejectionEnvelope envelope) {
         final Rejection rejection = envelope.getOuterObject();
         final ExternalMessage message = ExternalMessages.of(rejection, boundedContextName);
-        final ExternalMessageClass messageClass =
-                ExternalMessageClass.of(envelope.getMessageClass());
-        final Publisher channel = publisherHub.get(messageClass);
-        channel.publish(AnyPacker.pack(envelope.getId()), message);
 
-        return ImmutableSet.of(channel.toString());
+        final Iterable<Publisher> channels = router.route(message);
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/DomesticRejectionPublisher.java]: This portion of code looks a copy-paste changeset from `DomesticEventPublisher`.

File: server/src/main/java/io/spine/server/integration/route/Route.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+package io.spine.server.integration.route;
+
+import io.spine.server.integration.ChannelId;
+import io.spine.server.integration.ExternalMessage;
+import io.spine.server.integration.MessageChannel;
+import io.spine.server.integration.MessageRouted;
+
+/**
+ * The base interface for a route. The route is an entity that connects
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/route/Route.java]: Let's not use `entity` here. It has a slightly different meaning in Spine.

File: server/src/main/java/io/spine/server/integration/route/Route.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+ * @author Dmitry Ganzha
+ */
+public interface Route {
+
+    /**
+     * Checks if the message can be routed via this route.
+     *
+     * <p>A route can:
+     * <ul>
+     * <li>accept the message (by returning {@code MessageRouted} which will indicate that
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/route/Route.java]: Please reformat with indents. See similar cases in Spine.

File: server/src/main/java/io/spine/server/integration/route/RoutingSchema.java [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.integration.route;
+
+/**
+ * The base interface for a routing schema. Should be used with a {@linkplain Router message router}
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/route/RoutingSchema.java]: Please explain why you created it.

File: server/src/main/java/io/spine/server/integration/route/Router.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+
+package io.spine.server.integration.route;
+
+import io.spine.server.integration.ExternalMessage;
+import io.spine.server.integration.MessageChannel;
+import io.spine.server.integration.Publisher;
+
+/**
+ * The base interface for a message router. The {@code Router} is designed for routing
+ * messages to {@linkplain MessageChannel message channels} based on predefined
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/route/Router.java]: `predefined` conflicts with `dynamic router`.

File: server/src/main/java/io/spine/server/integration/route/action/ChannelAction.java [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+import io.spine.server.integration.route.Route;
+
+/**
+ * The base interface for message channel actions. The actions are used in
+ * {@linkplain Route#accept(ExternalMessage)} to verify if the message acceptable
+ * for the {@code Route}.
+ *
+ * @author Dmitry Ganzha
+ */
+public interface ChannelAction {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/route/action/ChannelAction.java]: Let's try to find another name. `ChannelAction.perform(channelID, myMessage)` call looks more than weird.

**Comment by dmitry-ganzha** [comment]
@armiol PTAL.

**Comment by armiol** [comment]
This feature is now included into #689.

**Pull request closed without merging**